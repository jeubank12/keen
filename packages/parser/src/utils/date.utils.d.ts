/// <reference types="react" />
/// <reference types="@emotion/core" />
import { TimePrecision } from '@keen.io/charts';
export declare const createLabelFormatter: (
  precision: TimePrecision
) => (label: import('react').ReactText) => import('react').ReactText;
