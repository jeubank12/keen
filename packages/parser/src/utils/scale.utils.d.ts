import { ScaleSettings } from '@keen.io/charts';
import { Query } from '../types';
export declare const createScaleSettings: ({
  interval,
}: Query) => Partial<ScaleSettings>;
