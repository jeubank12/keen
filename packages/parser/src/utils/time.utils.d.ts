import { TimePrecision } from '@keen.io/charts';
export declare const getPrecisionForInterval: (
  interval: string
) => TimePrecision;
