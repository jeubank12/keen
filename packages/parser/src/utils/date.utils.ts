import { timeFormat } from 'd3-time-format';
import { TimePrecision } from '@keen.io/charts';

const DATE_FORMATS: Record<TimePrecision, string> = {
  minute: '%I:%M:%S %p',
  hour: '%I:%M %p',
  day: '%b %a %y',
  week: '%d %a',
  month: '%b %y',
  year: '%b, %Y',
};

export const createLabelFormatter = (
  precision: TimePrecision
): ((label: string | number) => string | number) => {
  const format = DATE_FORMATS[precision] || DATE_FORMATS['month'];
  const formatTime = timeFormat(format);
  return (date: string | number) => formatTime(new Date(date));
};
