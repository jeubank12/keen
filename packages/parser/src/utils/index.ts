import { createScaleSettings } from './scale.utils';

export { createScaleSettings };
