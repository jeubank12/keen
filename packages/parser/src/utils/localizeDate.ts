import { Intervals } from '../types';
import { timeFormat } from 'd3-time-format';

const dateFormatter = timeFormat('%Y-%m-%dT%H:%M:%S');

const clearDateTime = (dateString: string) => {
  const date = new Date(dateString);
  date.setHours(0);
  date.setMinutes(0);
  return dateFormatter(date);
};

const DATE_FORMATS: Record<Intervals, any> = {
  daily: clearDateTime,
  hourly: (date: string) => date,
  minutely: (date: string) => date,
  weekly: clearDateTime,
  monthly: clearDateTime,
  yearly: clearDateTime,
};

/**
 * Provides time conversion to handle all dates
 * as localized browser's timezone by default
 *
 * @param date - date
 * @param interval - interval time precision
 * @return localized date
 *
 */
export const localizeDate = (date: string, interval: Intervals) =>
  DATE_FORMATS[interval](date);
