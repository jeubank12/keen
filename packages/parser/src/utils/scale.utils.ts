import { ScaleSettings } from '@keen.io/charts-utils';

import { getPrecisionForInterval } from './time.utils';
import { createLabelFormatter } from './date.utils';

export const createScaleSettings = (
  interval: string
): Partial<ScaleSettings> => {
  const precision = getPrecisionForInterval(interval);
  const formatLabel = createLabelFormatter(precision);

  return {
    precision,
    formatLabel,
  };
};
