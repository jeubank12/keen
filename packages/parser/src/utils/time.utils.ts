import { TimePrecision } from '@keen.io/charts';

const INTERVAL_TO_PRECISION: Record<string, TimePrecision> = {
  minutely: 'minute',
  hourly: 'hour',
  daily: 'day',
  weekly: 'week',
  monthly: 'month',
  yearly: 'year',
};

const CUSTOM_INTERVAL_TO_PRECISION: Record<string, TimePrecision> = {
  minutes: 'minute',
  hours: 'hour',
  days: 'day',
  weeks: 'week',
  months: 'month',
  years: 'year',
};

const isCustomInterval = (interval: string): boolean => {
  const regex = /every_[2-9]+\d*_(minutes|hours|days|weeks|months|years)/;
  return regex.test(interval);
};

const DEFAULT_INTERVAL = 'month';

export const getPrecisionForInterval = (interval: string): TimePrecision => {
  if (isCustomInterval(interval))
    return (
      CUSTOM_INTERVAL_TO_PRECISION[interval.split('_')[2]] || DEFAULT_INTERVAL
    );

  return INTERVAL_TO_PRECISION[interval] || DEFAULT_INTERVAL;
};
