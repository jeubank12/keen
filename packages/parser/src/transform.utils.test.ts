import { transformFromArray, transformFromNumber } from './transform.utils';

import { arrayValues } from './api.fixtures';

describe('@keen.io/parser - transform', () => {
  describe('transformFromNumber()', () => {
    it('should create structure for single "numeric" value', () => {
      const result = transformFromNumber(12);
      expect(result).toMatchSnapshot();
    });
  });

  describe('transformFromNumber()', () => {
    it('should create structure for multiple "collections" value', () => {
      const result = transformFromArray(arrayValues);
      expect(result).toMatchSnapshot();
    });
  });
});
