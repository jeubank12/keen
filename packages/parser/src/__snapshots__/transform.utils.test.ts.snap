// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@keen.io/parser - transform transformFromNumber() should create structure for multiple "collections" value 1`] = `
Object {
  "data": Object {
    "Africa": 1,
    "Cracow": 3,
    "Daegu": 12,
  },
  "keys": Set {
    "Africa",
    "Cracow",
    "Daegu",
  },
}
`;

exports[`@keen.io/parser - transform transformFromNumber() should create structure for single "numeric" value 1`] = `
Object {
  "formatLabel": undefined,
  "keys": Array [
    "value",
  ],
  "results": Array [
    Object {
      "name": "Result",
      "value": 12,
    },
  ],
}
`;
